lib/veres-one-did-doc.js

```diff
@@ -0,0 +1,456 @@
+
+    return {
+      '@context': constants.VERES_ONE_CONTEXT_URL,
+      target: this.id,
+      sequence: this.meta.sequence,
+      patch
+    };
+  }
+
+  /**
+   * Composes and returns a service id for a service name.
+   *
+   * @param serviceName {string}
+   *
+   * @returns {string} Service id
+   */
+  serviceIdFor(serviceName) {
+    if(!serviceName) {
+      throw new Error('Invalid service name.');
+    }
+
+    return this.id + ';service=' + encodeURIComponent(serviceName);
+  }
+
+  /**
+   * Finds a service endpoint in this did doc, given an id or a name.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   *
+   * @returns {object}
+   */
+  findService({id, name}) {
+    const jsonld = this.injector.use('jsonld');
+
+    const serviceId = id || this.serviceIdFor(name);
+
+    return jsonld
+      .getValues(this.doc, 'service')
+      .find(service => service.id === serviceId);
+  }
+
+  /**
+   * Tests whether this did doc has a service endpoint (by name or id).
+   * One of `id` or `name` is required.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   *
+   * @returns {boolean}
+   */
+  hasService({id, name}) {
+    return !!this.findService({id, name});
+  }
+
+  /**
+   * Adds a service endpoint to this did doc.
+   * One of `id` or `name` is required.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   * @param type {string} Endpoint type (e.g. 'AgentService')
+   * @param endpoint {string} Endpoint uri (e.g. 'https://agent.example.com')
+   * @param [options] {object} Any additional properties of endpoint
+   */
+  addService({id, name, type, endpoint, ...options}) {
+    const jsonld = this.injector.use('jsonld');
+
+    const serviceId = id || this.serviceIdFor(name);
+
+    if(!type) {
+      throw new Error('Service endpoint type is required.');
+    }
+    if(!endpoint) {
+      throw new Error('Service endpoint uri is required.');
+    }
+
+    if(this.findService({id, name})) {
+      throw new Error('Service with that name or id already exists.');
+    }
+
+    jsonld.addValue(this.doc, 'service', {
+      id: serviceId,
+      serviceEndpoint: endpoint,
+      type,
+      ...options
+    }, {
+      propertyIsArray: true
+    });
+  }
+
+  /**
+   * Removes a service endpoint from this did doc.
+   * One of `id` or `name` is required.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   */
+  removeService({id, name}) {
+    const jsonld = this.injector.use('jsonld');
+
+    const serviceId = id || this.serviceIdFor(name);
+
+    const services = jsonld
+      .getValues(this.doc, 'service')
+      .filter(service => service.id !== serviceId);
+    if(services.length === 0) {
+      jsonld.removeProperty(this.doc, 'service');
+    } else {
+      this.doc.service = services;
+    }
+  }
+
+  addKey({key, suiteId, owner = this.id}) {
+    // Add public key node to the DID Doc
+    const suite = this.suiteForId(suiteId);
+    if(!suite) {
```

Comment:
Why not have `suiteForId` throw the error, then any consumer of that API does not have to perform this type of check?

Reply:
So, I was thinking - not finding a suite for an ID is not necessarily an error condition. So it's better left up to consuming code to decide?